verifier/src/lib.rs

```diff
@@ -3,10 +3,222 @@
 #![doc = include_str!("../README.md")]
 #![deny(missing_docs, missing_debug_implementations, unsafe_code)]
 
+use std::fmt::{Debug, Formatter};
+
+type Result<T> = std::result::Result<T, VerificationError>;
+
+#[derive(Debug, Eq, PartialEq)]
+/// Failed to verify: {0}.
+pub struct VerificationError(String);
+
+impl<S: Into<String>> From<S> for VerificationError {
+    fn from(message: S) -> Self {
+        Self(message.into())
+    }
+}
+
+/// A verification step. These can chained together using the [`Or`] and [`And`]
+/// types.
+pub trait VerificationStep {
+    /// Performs a verification operation for the [`VerificationStep`].
+    ///
+    /// When verification fails the [`VerificationError`] should contain a
+    /// message communicating the cause of the failure.
+    fn verify(&self) -> Result<()>;
+}
+
+impl Debug for dyn VerificationStep {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("VerificationStep").finish()
+    }
+}
+
+/// Will perform a logical and operation for the [`VerificationStep::verify()`]
+/// operation.
+///
+/// This is will be a short circuiting operation. If the `left` side fails
+/// the `right` side will *not* be exercised.
+#[derive(Debug)]
+pub struct And<L, R> {
+    left: L,
+    right: R,
+}
+
+impl<L: VerificationStep, R: VerificationStep> And<L, R> {
+    /// Create a new [`And`] instance
+    ///
+    /// # Arguments:
+    /// * `left` - The left, or first, [`VerificationStep`] to perform. If this
+    ///    fails the `right` will not be attempted.
+    /// * `right` - The right, or second, [`VerificationStep`] to perform.
+    pub fn new(left: L, right: R) -> Self {
+        Self { left, right }
+    }
+
+    /// The left side of this logical and instance.
+    pub fn left(&self) -> &L {
+        &self.left
+    }
+
+    /// The right side of this logical and instance.
+    pub fn right(&self) -> &R {
+        &self.right
+    }
+}
+
+impl<L: VerificationStep, R: VerificationStep> VerificationStep for And<L, R> {
+    fn verify(&self) -> Result<()> {
+        self.left.verify()?;
+        self.right.verify()
```

Comment:
`And` should evaluate both sides and return the union of their errors (i.e. the log output should be able to say _everything_ that's wrong with the evidence).

Reply:
I don't think I'm fully following this. From the comment above:
> This also means we can never communicate securely between a hardware element that doesn't support alloc and an enclave.

It seems like we need some kind of growing structure to be able to hold all of the errors as they percolate up, or we need to say we can hold a fixed size of errors.  But either way we need some way to hold indicators that "TCBInfo out of date" and "Debug flag was on".  I can understand if we don't want the informative messages/info of "SVN version was 3, needs to be at least 5", or maybe we can choose some errors to contain smaller concrete values.

Reply:
using CtOption with the self::error logic